src/foam/nanos/auth/CachingAuthService.java

```diff
@@ -82,59 +98,58 @@ public boolean checkUser(foam.core.X x, foam.nanos.auth.User user, String permis
     return permissionCheck;
   }
 
-  public void purgeCache(X x) {
-    User user = getUserFromContext(x);
-    if ( user != null ) {
-      userPermissionCache_.remove(user.getId());
+  public void purgeCache(Object obj) {
+    Long userId = null;
+
+    // Check for supported types to purge single user permission map
+    if ( obj instanceof User ) {
+      userId = ((User) obj).getId();
+    } else if ( obj instanceof UserCapabilityJunction ) {
+      userId = ((UserCapabilityJunction) obj).getSourceId();
+    }
+
+    if ( userId != null ) {
+      // Reset single user permission map
+      userPermissionCache_.remove(userId);
+    } else {
+      // Reset permission cache
+      // Reassigning LRU cache will garbage collect old cache and reset to empty cache with initial capacity
+      userPermissionCache_ = new LRULinkedHashMap<>(CACHE_NAME, CACHE_SIZE);
```

Comment:
Better to call .clear() because this isn't thread-safe. Variables updated outside of a synchronized block are not guaranteed to be seen by other threads.

Reply:
I can call clear(), however, there is no way to reduce the size of a hashmap ... so if we get max capacity (10000) and we invalidate the cache, the permission map will forever have 10000 elements using clear().  

It seemed like you preferred a smaller cache if appropriate when we chatted about initial capacity?

Reply:
Changed to use clear() in https://github.com/kgrgreer/foam3/pull/857